File: nim/day13.nim
import strutils, math

proc parse(path: string): (int, seq[(int, int)]) =
  let file = open path
  result[0] = parseInt readLine(file)
  let buses = readLine(file).split(',')
  for i, bus in buses:
    if bus != "x":
      result[1].add (i, parseInt(bus))

func part1(timestamp: int, buses: seq[(int, int)]): int =
  var earliest = int.high
  for (_, bus) in buses:
    let wait = bus - (timestamp mod bus)
    if wait < earliest:
      earliest = wait
      result = bus * wait

func part2(buses: seq[(int, int)]): int =
  var prime = 1
  for (i, bus) in buses:
    let delay = floorMod(bus-i, bus)
    let newPrime = prime * bus
    for n in countup(result, newPrime, prime):
      if n mod bus == delay:
        result = n
        prime = newPrime
        break

let (timestamp, buses) = parse "inputs/13.txt"
echo part1(timestamp, buses)
echo part2(buses)

File: python/day13.py
from functools import reduce
from operator import mul


timestamp, buses = (int((f := open("inputs/13.txt")).readline().strip()),
                    [(i, int(bus)) for i, bus in enumerate(f.readline().strip().split(',')) if bus != 'x'])

crt = (lambda prev, curr:
          (lambda res, prime, i, bus:
              next((n, prime*bus) for n in range(res, prime*bus, prime)
                                  if not (n+i) % bus))
          (*prev, *curr))


print(mul(*(reduce(min, ((-timestamp % bus, bus) for _, bus in buses)))))
print(reduce(crt, buses, (0, 1))[0])
